File: Tests/AINotebookCoreTests/NotebookStoreTransformationsTests.swift
import XCTest
@testable import AINotebookCore

@MainActor
final class NotebookStoreTransformationsTests: XCTestCase {

    func testCreateAndListTransformations() throws {
        let store = try NotebookStore(path: .inMemory)
        _ = try store.createTransformation(
            name: "CustomA", promptTemplate: "Summarize:\n{{source_text}}", scope: .source, isBuiltin: false
        )
        _ = try store.createTransformation(
            name: "CustomB", promptTemplate: "Do X", scope: .source, isBuiltin: false
        )
        let nonBuiltin = try store.transformations().filter { !$0.isBuiltin }
        XCTAssertEqual(nonBuiltin.count, 2)
    }

    func testUpdateAndDeleteCustomTransformation() throws {
        let store = try NotebookStore(path: .inMemory)
        let t = try store.createTransformation(
            name: "C", promptTemplate: "old", scope: .source, isBuiltin: false
        )
        try store.updateTransformation(id: t.id!, name: "C2", promptTemplate: "new")
        let reloaded = try XCTUnwrap(store.transformations().first { $0.id == t.id })
        XCTAssertEqual(reloaded.name, "C2")
        XCTAssertEqual(reloaded.promptTemplate, "new")
        try store.deleteTransformation(id: t.id!)
        XCTAssertEqual(try store.transformations().filter { !$0.isBuiltin }.count, 0)
    }

    func testRecordRunCreatesRow() throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")
        let s  = try store.createSource(notebookId: nb.id!, type: .text, title: "x", uri: nil, rawPath: nil)
        let t  = try store.createTransformation(name: "T", promptTemplate: "p", scope: .source, isBuiltin: false)
        let n  = try store.createNote(notebookId: nb.id!, title: "T result", bodyMd: "x")
        let run = try store.recordTransformationRun(
            transformationId: t.id!, sourceId: s.id!, resultNoteId: n.id!
        )
        XCTAssertNotNil(run.id)
        let runs = try store.transformationRuns()
        XCTAssertEqual(runs.count, 1)
        XCTAssertEqual(runs[0].resultNoteId, n.id)
    }
}

File: windows/tests/AINotebook.Core.Tests/Storage/NotebookStoreTransformationsTests.cs
using AINotebook.Core.Models;
using AINotebook.Core.Storage;
using Xunit;

namespace AINotebook.Core.Tests.Storage;

public class NotebookStoreTransformationsTests
{
    [Fact]
    public void CreateListUpdateDelete()
    {
        using var store = new NotebookStore(StorePath.InMemory);
        var t = store.CreateTransformation("Custom", "do {{source_text}}", TransformationScope.Source, isBuiltin: false);
        Assert.Contains(store.Transformations(), x => x.Id == t.Id && !x.IsBuiltin);

        store.UpdateTransformation(t.Id!.Value, "Custom2", "redo {{source_text}}", "desc");
        var updated = store.Transformations().Single(x => x.Id == t.Id);
        Assert.Equal("Custom2", updated.Name);
        Assert.Equal("redo {{source_text}}", updated.PromptTemplate);

        store.DeleteTransformation(t.Id!.Value);
        Assert.DoesNotContain(store.Transformations(), x => x.Id == t.Id);
    }

    [Fact]
    public void OrderedBuiltinsFirstThenNameAsc()
    {
        using var store = new NotebookStore(StorePath.InMemory);
        store.CreateTransformation("aaa-custom", "{{source_text}}", TransformationScope.Source, isBuiltin: false);
        var list = store.Transformations();
        // all builtins precede all non-builtins
        int firstNonBuiltin = list.ToList().FindIndex(x => !x.IsBuiltin);
        Assert.True(list.Take(firstNonBuiltin).All(x => x.IsBuiltin));
    }

    [Fact]
    public void RecordRunAndList()
    {
        using var store = new NotebookStore(StorePath.InMemory);
        var t = store.Transformations().First(x => x.IsBuiltin);
        var nb = store.CreateNotebook("N").Id!.Value;
        var note = store.CreateNote(nb, "out", "body");
        var run = store.RecordTransformationRun(t.Id!.Value, sourceId: null, resultNoteId: note.Id);
        Assert.Equal(note.Id, run.ResultNoteId);
        Assert.Contains(store.TransformationRuns(), r => r.Id == run.Id);
    }
}
